### ISSUE TITLE
Support raw javadocs

### ISSUE THREAD

**Issue by bergerkiller** [dzikoysk/reposilite#1729]
### What happened?

When viewing the javadoc of an artifact, the true javadoc page is displayed inside an IFRAME. This is presumably done so that Reposilite can put a banner above the javadocs.

~~The problem is that Oracle doesn't support this at all. The IFRAME breaks the search functionality of the javadoc because it tries to navigate the main window, which the page in the IFRAME has no access to (CORS).~~ **Fixed by building the javadoc using JDK19**

IDE's which support html-importing of javadocs don't work with it because it expects a /index.html with the page contents. People can't copy the direct link to a relevant javadoc page to save for later or share with colleagues.

The current workaround is to visit the page, and then append _.cache/unpack/index.html_ to the URL.

Bluntly put - Is it possible to turn this off? Ideally it would just put the javadocs directly at the URL, instead of a container page.

Weirdly, it seems javadocs including search do work for [the reposilite project itself](https://maven.reposilite.com/javadoc/releases/com/reposilite/reposilite/latest/index-all.html) but I cant reproduce this behavior on my end.

### Reposilite version

3.x

**Comment by dzikoysk** [comment]
This is known Java bug that has been fixed lately in javadoc tool:

* https://bugs.openjdk.org/browse/JDK-8268866

Custom index.html file was originally proposed by the old plugin, so we didn't change that, but we've left a possibility to request unpacked javadocs content just like you've said. We're using iframe for 2 reasons:

* Add custom navigation header, so we can make navigation and extend it in the future with e.g. download link/version switcher
* Embed javadocs [in sandbox](https://github.com/dzikoysk/reposilite/blob/main/reposilite-backend/src/main/kotlin/com/reposilite/javadocs/JavadocView.kt#L94) - there's a possibility that javadocs jar may contain malicious js script from e.g. mirrored repository. Of course it's not perfect, but provides basic protection for things like access to parent's localStorage etc.

**Comment by bergerkiller** [comment]
Thanks, that does make a lot of sense. Im surprised it was fixed, because I picked the latest javadoc maven plugin version and the problem was still there. Then searched a bit and saw a bunch of oracle issue tickets relating to IFRAMES that were closed with 'WONTFIX'. Or is it not the javadoc maven plugin taht needs a version bump, but something else?

In that case I have two requests:
- Some sort of direct link to this cache location on the javadoc page for people that are trying to import it into their IDE. Personally I just use the javadoc jar but apparently there is a need for this functionality.
- Alternative to above: Why not put the true javadoc at /index.html and have / do the iframe? It **seems** that this is how the reposilite javadocs are hosted [as this link also works](https://maven.reposilite.com/javadoc/releases/com/reposilite/reposilite/latest/)
- Some sort of #ref api in the page URL bar that shows where the IFRAME is navigated. But I'm afraid this isn't possible because reading of the IFRAME url is blocked by default for security reasons in browsers

**Comment by bergerkiller** [comment]
Happy to report rebuilding the projects on Java 19 fixed the issues :+1:

**Comment by dzikoysk** [comment]
Nice! Maybe we could do sth like:

* Keep `https://maven.reposilite.com/javadoc/releases/com/reposilite/reposilite/latest/` & add link to the raw version in header
* `https://maven.reposilite.com/javadoc-raw/releases/com/reposilite/reposilite/latest/` that'd be just the 1:1 jar content like with `.cache/unpack/[...]`

I think it could be relatively convenient for both cases.

**Comment by bergerkiller** [comment]
I'm all for that. One problem I ran into was that the .cache directory isn't generated until the main / is called at least once. It also seems to have some background process to automatically remove it, maybe when a new snapshot is pushed. So in addition, it should generate the /javadoc-raw/ content automatically when a page under that url is first requested

**Comment by dzikoysk** [comment]
> I'm all for that. One problem I ran into was that the .cache directory isn't generated until the main / is called at least once. It also seems to have some background process to automatically remove it, maybe when a new snapshot is pushed. So in addition, it should generate the /javadoc-raw/ content automatically when a page under that url is first requested

This is required to obtain gav from request uri, so we can find related javadoc file. During first request for e.g. `https://maven.reposilite.com/javadoc/releases/com/reposilite/reposilite/latest/a/b/c/d/file.html` you're unable to say what's the groupid/artifactid/version.

> IMO, having /latest/raw/index.html or /latest/unpacked/index.html might be a little cleaner though. Especially because to get there one first has to visit the /javadoc/ url anyway. (Unless you plan to add a second book icon in the repository for raw read? But dunno.)

The only problem with such approach is that it won't work if someone will have artifact called e.g. `raw`. It might be dumb to care about such niche, but this is something that we should be aware of. That's why in current impl, the location of this internal directory starts with `.`

**Comment by bergerkiller** [comment]
What about /?raw=true?

**Comment by dzikoysk** [comment]
Would you like to try to submit PR with such enhancement? :)

**Comment by bergerkiller** [comment]
When I have the time for it I might. I'm a java developer primarily so the kotlin stuff is a bit new to me. Good enough at it to copy-paste stuff around and make tweaks, but not confident enough in it to redesign the network routing stuff.

For stability sake, probably better someone more knowledgeable picks that up. With search fixed, my primary concern is resolved so it has gotten less important

**Comment by dzikoysk** [comment]
Sure, but just in case, it's open source and you can learn here, so if you'd decide someday to commit bigger changes in Kotlin to the repo, I'll try to guide you and highlight things we could improve :)

In the meantime, @Rollczi would you like to implement this?

**Comment by Rollczi** [comment]
I'm busy for a month. 😿
